### ISSUE TITLE
impl Into<B>.into() has type {unknown}

### ISSUE THREAD

**Issue by Geo25rey** [rust-lang/rust-analyzer#15521]
**rust-analyzer version**: 0.3.1623-standalone (2fbe69d11 2023-08-12)

**rustc version**: rustc 1.72.0 (5680fa18f 2023-08-23)

**relevant settings**: all default settings on stable version channel

### Description
In the `main` function below, `let b` should be annotated with `: B` as the rust compiler understands, but instead generates no annotation and when highlighted has type `{unknown}`. Additionally, on the following line, `b.a` does not generate an error, does not provide any type information, and has white syntax highlighting on the `a` instead of the blue that you get when manually adding the `: B` annotation on the previous line.

#### Code to Reproduce
```rust
pub struct B {
    a: u64,
}

#[derive(Clone)]
pub struct A(u64);

impl Into<B> for A {
    fn into(self) -> B {
        B { a: self.0 }
    }
}

fn main() {
    let a: A = A(8);
    let b = a_to_b(&a).into(); // this is type B but rust-analyzer says the type is {unknown}
    println!("b.a = {}", b.a)
}

fn a_to_b(b_compat: &(impl Into<B> + Clone)) -> impl Into<B> {
    b_compat.clone()
}
```

**Comment by laundmo** [comment]
a much easier reproduction of what i belive to be the same bug
```rs
fn impl_into(arg: impl Into<f64>){
    let a = arg.into();
    //  ^ this fails to resolve, shows {unknown}
    dbg!(a);
}

fn main() {
    impl_into(1.1f32);    
}
```

**Comment by HKalbasi** [comment]
Duplicate of #5514
